dmoj/executors/BF.py

```diff
@@ -1,80 +1,215 @@
+  %size0 = getelementptr i8*, i8** %argv, i64 1
+  %size1 = load i8*, i8** %size0
+  %size = call i64 @strtoll(i8* %size1, i8** null, i32 10)
+  ; mmap NULL size PROT_READ|PROT_WRITE MAP_PRIVATE|MAP_ANONYMOUS -1 0
+  %p = call i8* @mmap(i8* null, i64 %size, i32 3, i32 34, i32 -1, i64 0)
+  %pi = ptrtoint i8* %p to i64
+  %pfail = icmp uge i64 %pi, -4095  ; -4095 to -1 indicates error
+  br i1 %pfail, label %end2, label %bf
+
+end2:
+  ret i32 2
+
+bf:
+  {code}
+  ret i32 0
 }
 """
 
-trans = {
-    ord('>'): b'++p;',
-    ord('<'): b'--p;',
-    ord('+'): b'++*p;',
-    ord('-'): b'--*p;',
-    ord('.'): b'putchar(*p);',
-    ord(','): b'*p=getchar();',
-    ord('['): b'while(*p){',
-    ord(']'): b'}',
-}
 
+# Helper for simple loops. Consider the loop "[>>++>-<<<-]".
+# This helper translates it into {2: 2, 3: -1}. Later, the LLVM looks like this:
+#   if (p[0] != 0) { p[2] += p[0] * 2; p[3] += p[0] * -1; p[0] = 0; }
+# The if-guard is needed so that programs like "[-<+>]" don't segfault.
+def simple_loop(bf_inst: List[Union[str, Dict[int, int]]]) -> Optional[Dict[int, int]]:
+    loop: Dict[int, int] = {}
+    ptr = 0
+    for c in bf_inst:
+        assert c in ['>', '<', '+', '-']
+        if c == '>':
+            ptr += 1
+        elif c == '<':
+            ptr -= 1
+        elif c == '+':
+            loop[ptr] = loop.get(ptr, 0) + 1
+        elif c == '-':
+            loop[ptr] = loop.get(ptr, 0) - 1
+    if ptr != 0:
+        return None
+    if loop.get(0) != -1:
+        return None
+    loop.pop(0)
+    return loop
 
-class Executor(CExecutor):
-    test_program = ',+[-.,+]'
 
-    def __init__(self, problem_id, source_code, **kwargs):
-        if self._has_invalid_brackets(source_code):
-            raise CompileError(b'Unmatched brackets\n')
-        code = template.replace(b'{code}', b''.join(map(trans.get, source_code, itertools.repeat(b''))))
-        super().__init__(problem_id, code, **kwargs)
+def compile_to_llvm(source_code: bytes) -> bytes:
+    # Do a pass to optimize simple loops.
+    # This speeds up the opt step, because opt takes a long time to optimize simple loops.
+    bf_inst: List[Union[str, Dict[int, int]]] = []
+    bf_stack: List[int] = []
+    simple = True  # Indicates if current loop only uses ><+-
+    for b in source_code:
+        if b in b'><+-':
+            bf_inst.append(chr(b))
+        elif b in b',.':
+            bf_inst.append(chr(b))
+            simple = False
+        elif b == b'['[0]:
+            bf_inst.append('[')
+            bf_stack.append(len(bf_inst))
+            simple = True
+        elif b == b']'[0]:
+            if not bf_stack:
+                raise CompileError(b'Unmatched brackets\n')
+            i = bf_stack.pop()
+            loop = simple_loop(bf_inst[i:]) if simple else None
+            if loop is None:
+                bf_inst.append(']')
+            else:
+                while len(bf_inst) >= i:
+                    bf_inst.pop()
```

Comment:
Why not just `bf_inst = bf_inst[:i+1]` or something like that?

Reply:
This loop is O(popped items). `bf_inst = bf_inst[:i+1]` is O(i) and will make compilation O(N^2).

Reply:
Have you tried `bf_inst[i+1:] = []` then?

Reply:
Done. `bf_inst[i - 1 :] = [loop]` runs fast.